### PR TITLE
[SelectionDAG] Make ARITH_FENCE support half and bfloat type

### PR DESCRIPTION
**Pull request by phoebewang**

### CODE REVIEW

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll [hunk 1]
```diff
@@ -157,6 +157,91 @@ define <8 x float> @f6(<8 x float> %a) {
...
+; X64-NEXT:    shll $16, %eax
+; X64-NEXT:    movd %eax, %xmm0
+; X64-NEXT:    addss %xmm0, %xmm0
+; X64-NEXT:    callq __truncsfbf2@PLT
+; X64-NEXT:    popq %rax
+; X64-NEXT:    retq
+  %1 = fadd fast bfloat %a, %a
+  %t = call bfloat @llvm.arithmetic.fence.bf16(bfloat %1)
+  %2 = fadd fast bfloat %a, %a
+  %3 = fadd fast bfloat %1, %2
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/arithmetic_fence2.ll:239]: Use named values in tests. Also drop the flags

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll [hunk 2]
```diff
@@ -157,6 +157,91 @@ define <8 x float> @f6(<8 x float> %a) {
...
+  ret bfloat %3
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/arithmetic_fence2.ll:242]: Test vectors too?

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll (hunk 1 above)
Comment (phoebewang) [reply llvm/test/CodeGen/X86/arithmetic_fence2.ll]: Done.

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll (hunk 2 above)
Comment (phoebewang) [reply llvm/test/CodeGen/X86/arithmetic_fence2.ll:498]: Done.

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/X86/arithmetic_fence2.ll]: Actually can drop all the instructions that aren't just the fence

Comment (phoebewang) [reply llvm/test/CodeGen/X86/arithmetic_fence2.ll]: Done.

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll [hunk 3]
```diff
@@ -157,6 +157,119 @@ define <8 x float> @f6(<8 x float> %a) {
...
+; X64-NEXT:    movdqa %xmm0, %xmm1
+; X64-NEXT:    psrld $16, %xmm1
+; X64-NEXT:    #ARITH_FENCE
+; X64-NEXT:    #ARITH_FENCE
+; X64-NEXT:    punpcklwd {{.*#+}} xmm0 = xmm0[0],xmm1[0],xmm0[1],xmm1[1],xmm0[2],xmm1[2],xmm0[3],xmm1[3]
+; X64-NEXT:    retq
+  %b = call <2 x half> @llvm.arithmetic.fence.v2f16(<2 x half> %a)
+  ret <2 x half> %b
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/arithmetic_fence2.ll:214]: You skipped 3 x, the case most likely to break

**Review by arsenm — APPROVED** [review]
LGTM but should also test the 3x case

File: llvm/test/CodeGen/X86/arithmetic_fence2.ll (hunk 3 above)
Comment (phoebewang) [reply llvm/test/CodeGen/X86/arithmetic_fence2.ll:214]: Done.

**Pull request merged**